File: src/lib/motion.ts
/**
 * Pulse premium motion system (R22) — single source of truth for springs,
 * easings and the app-wide particle FX contract. Every surface imports from
 * here so the whole app moves with one physical language.
 *
 * Principles (from the Telegram/iMessage/Discord motion research):
 * - Springs over durations: mass/stiffness/damping, velocity-aware.
 * - Non-linear cubic-bezier easings for anything that isn't a spring.
 * - Compositor-only properties (transform/opacity) + will-change isolation.
 * - Squash-and-stretch: fast enters overshoot, exits are quicker than enters.
 */
import type { Transition } from 'framer-motion'

/** Spring presets — pick by intent, never hand-roll one-off transitions. */
export const spring = {
  /** UI chrome: tabs, pills, indicators. Fast settle, no wobble. */
  snappy: { type: 'spring', stiffness: 500, damping: 34, mass: 0.9 },
  /** Sheets, cards, layout shifts. A touch of life without bounce. */
  soft: { type: 'spring', stiffness: 300, damping: 28, mass: 1 },
  /** Playful pops: badges, reactions, bubbles. Visible overshoot. */
  bouncy: { type: 'spring', stiffness: 620, damping: 20, mass: 0.85 },
  /** Ambient drift: parallax, hero glow. Barely-there motion. */
  gentle: { type: 'spring', stiffness: 140, damping: 22, mass: 1.1 },
} satisfies Record<string, Transition>

/** Signature cubic-beziers for non-spring (opacity/blur/color) transitions. */
export const ease = {
  /** "Swift out" — content appearing. */
  out: [0.16, 1, 0.3, 1] as [number, number, number, number],
  /** Balanced in-out for crossfades/morphs. */
  inOut: [0.65, 0, 0.35, 1] as [number, number, number, number],
  /** Back-overshoot for stickers/emoji-scale pops. */
  overshoot: [0.34, 1.56, 0.64, 1] as [number, number, number, number],
}

/** Standard stagger recipe for list/tile entrances. */
export const stagger = (i: number, step = 0.04, cap = 10) =>
  Math.min(i, cap) * step

/** Press micro-interaction — pair with `whileTap` for tactile buttons. */
export const pressTap = { scale: 0.94 }
export const pressSpring = spring.snappy

// ── App-wide particle FX contract (decoupled via a window event so any
// surface can fire bursts without importing the FX layer) ────────────────

export type ParticleKind = 'confetti' | 'hearts' | 'stars' | 'burst'

export interface ParticleBurstDetail {
  /** viewport x 0..1 (default 0.5) */
  x?: number
  /** viewport y 0..1 (default 0.6) */
  y?: number
  kind?: ParticleKind
  /** particle count (default 80) */
  count?: number
}

export const PARTICLE_BURST_EVENT = 'pulse:particle-burst'

/** Fire a full-screen particle celebration from anywhere in the app. */
export function fireParticles(detail: ParticleBurstDetail = {}): void {
  if (typeof window === 'undefined') return
  window.dispatchEvent(new CustomEvent<ParticleBurstDetail>(PARTICLE_BURST_EVENT, { detail }))
}

/** Reduced-motion respect for every crew to reuse. */
export function prefersReducedMotion(): boolean {
  return typeof window !== 'undefined' &&
    window.matchMedia?.('(prefers-reduced-motion: reduce)').matches === true
}

File: apps/android/ui/src/main/kotlin/app/pulse/ui/Motion.kt
package app.pulse.ui

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.graphics.graphicsLayer

/**
 * Pulse motion tokens — the native (Compose physics) translation of the web
 * motion system (src/lib/motion.ts). Same physical language: springs over
 * durations, squash-and-stretch, faster exits than enters.
 *
 * Web presets (stiffness/damping/mass) → Compose (stiffness, dampingRatio):
 *   snappy 500/34/0.9 → 560 / 0.85   (chrome, tabs, press)
 *   soft   300/28/1.0 → 300 / 0.82   (sheets, cards, layout)
 *   bouncy 620/20/0.85 → 730 / 0.44  (badges, reactions, bubbles)
 *   gentle 140/22/1.1 → 120 / 0.90   (ambient drift, parallax)
 */
object PulseMotion {
    fun <T> snappy() = spring<T>(dampingRatio = 0.85f, stiffness = 560f)
    fun <T> soft() = spring<T>(dampingRatio = 0.82f, stiffness = 300f)
    fun <T> bouncy() = spring<T>(dampingRatio = 0.44f, stiffness = 730f)
    fun <T> gentle() = spring<T>(dampingRatio = 0.9f, stiffness = 120f)

    const val PRESS_SCALE = 0.94f
}

/**
 * Tactile press micro-interaction (web `whileTap: { scale: 0.94 }`).
 * Pair with a haptic at the interaction site.
 */
fun Modifier.pulsePress(
    interactionSource: MutableInteractionSource,
    enabled: Boolean = true,
): Modifier = composed {
    val pressed by interactionSource.collectIsPressedAsState()
    val scale by animateFloatAsState(
        targetValue = if (pressed && enabled) PulseMotion.PRESS_SCALE else 1f,
        animationSpec = PulseMotion.snappy(),
        label = "pulsePress",
    )
    graphicsLayer {
        scaleX = scale
        scaleY = scale
    }
}

@Composable
inline fun rememberPressSource(): MutableInteractionSource = remember { MutableInteractionSource() }
